src/targets/pubDev.ts

```diff
@@ -0,0 +1,208 @@
+ */
+export type PubDevTargetConfig = PubDevTargetOptions &
+  Record<SecretsType, string>;
+
+type PartialTargetConfig = { name: string; value: string | undefined }[];
+
+/**
+ * Target responsible for uploading files to pub.dev.
+ */
+export class PubDevTarget extends BaseTarget {
+  /** Target name */
+  public readonly name: string = 'pub-dev';
+  /** Target options */
+  public readonly pubDevConfig: PubDevTargetConfig;
+  /** GitHub repo configuration */
+  public readonly githubRepo: GitHubGlobalConfig;
+
+  public constructor(
+    config: TargetConfig,
+    artifactProvider: BaseArtifactProvider,
+    githubRepo: GitHubGlobalConfig
+  ) {
+    super(config, artifactProvider, githubRepo);
+    this.pubDevConfig = this.getPubDevConfig();
+    this.checkRequiredSoftware();
+    this.githubRepo = githubRepo;
+  }
+
+  /**
+   * Returns the pub-dev config with the required data (e.g. environment
+   * variables) for this target. If there's a configuration requirement missing,
+   * raises an error.
+   *
+   * @returns the pub-dev config for this target.
+   */
+  private getPubDevConfig(): PubDevTargetConfig {
+    // We could do `...this.config`, but `packages` is in a list, not array format in `.yml`
+    // so I wanted to keep setting the defaults unified.
+    return {
+      dartCliPath: this.config.dartCliPath || 'dart',
+      packages: this.config.packages
+        ? Object.keys(this.config.packages)
+        : ['.'],
+      ...this.getTargetSecrets(),
+    };
+  }
+
+  private getTargetSecrets(): Record<SecretsType, string> {
+    const secrets = targetSecrets.map(name => {
+      checkEnvForPrerequisite({ name });
+      return {
+        name,
+        value: process.env[name],
+      };
+    });
+    return this.reduceConfig(secrets);
+  }
+
+  private reduceConfig(config: PartialTargetConfig): Record<string, string> {
+    return config.reduce((prev, current) => {
+      return {
+        ...prev,
+        [current.name]: current.value,
+      };
+    }, {});
+  }
+
+  /**
+   * Checks whether the required software to run this target is available
+   * in the system. It assumes the config for this target to be available.
+   * If there's required software missing, raises an error.
+   */
+  private checkRequiredSoftware(): void {
+    this.logger.debug(
+      'Checking if Dart CLI is available: ',
+      this.pubDevConfig.dartCliPath
+    );
+    checkExecutableIsPresent(this.pubDevConfig.dartCliPath);
+  }
+
+  /**
+   * Uploads all files to pub.dev using Dart CLI.
+   *
+   * @param version New version to be released
+   * @param revision Git commit SHA to be published
+   */
+  public async publish(_version: string, revision: string): Promise<any> {
+    await this.createCredentialsFile();
+
+    await withTempDir(
+      async directory => {
+        await this.cloneRepository(this.githubRepo, revision, directory);
+        await forEachChained(this.pubDevConfig.packages, async pkg =>
```

Comment:
`forEachChained`:

> This results in a chain of asynchronous actions that resolves once the last item action has completed. In contrast, `Promise.all` executes each promise simultaneously.

Do we need to publish packages one by one, or can we publish multiple packages at the same time? Why are we using `forEachChained` over `Promise.all` (or any other `Promise` method that allows us to publish multiple packages at the same time)?

Reply:
Publishing serially does have the added benefit of making it easier to debug.

Reply:
This + we use `showStdout` in `spawnProcess`, which would be useless with concurrent uploads, as all output lines would render in random order.